### ISSUE TITLE
Layer-related suggestions

### ISSUE THREAD

**Issue by girlwithglasses** [kotojs/kotojs#7]
I'm finding it very useful to be able to subclass the `Layer` class to create layers with specialised functionality that can be used by various different charts (e.g. a tooltip layer). At the moment, I'm making my own build with a slightly customised `Layer` object that acts as a base for the various subclasses. These alterations may be useful to others, so I thought I would mention them here, too.

- expose the `Layer` constructor by default to make it easy to subclass

- use duck-typing to identify Layer objects that are being added to a chart: I add my custom layer objects by calling

        chart.layer( 'name', new FunkyLayer( selection, layer_args ) );

thus avoiding the default Layer constructor. The chart tests for `selection instanceof Layer` before adding the layer, which fails for custom layer objects. Locally, I've edited chart.js to test whether the layer object has `dataBind` and `insert` (the two mandatory Layer functions) defined, i.e.:

    // we are reattaching a previous layer, which the
    // selection argument is now set to.
    if (arguments.length === 2) {

      if (selection instanceof Layer || ( 'undefined' !== typeof selection.dataBind && 'undefined' !== typeof selection.insert ) ) {
        selection._chart = this;
        this._layers.set(name, selection);
        return this._layers.get(name);
      } else {
        kotoAssert(false, 'When reattaching a layer, the second argument must be a koto layer');
      }
    }

- the `dataBind` function doesn't seem to have a way to access the parent layer object; this could be enabled with the following edit to `Layer.draw`:

old:

    bound = this.dataBind.call(this._base, data);

new:

    bound = this.dataBind.call(this._base, data, this);

(note the `this` argument has been added)

**Comment by nicksrandall** [comment]
@girlwithglasses Thanks for the feedback!

I like the idea of exposing the Layer class so that it can be subclassed and reused for common components. I'm thinking I'll store it as a static property on the `Koto` namespace so that people can use it like: 

```js
class FunkyLayer extends Koto.Layer {
  // override or extend methods here
}
```

Correct me if I'm wrong but it sounds like that instead of subclassing `Layer` your are mocking Layer with a new class that has the same required methods as Layer and thus you needed to add the extra conditions on the `instanceof` check mentioned above. Is there a reason that you're mocking Layer and not just extending it?

I don't have a problem with passing the context to the dataBind method as you described above.

## TO RECAP:
I'm going to add the context as a second parameter to the dataBind method and I'm going to globally expose the Layer constructor (class) by attaching it to the `Koto` name space as `Koto.Layer`;

Do you have any problems with that?

**Comment by girlwithglasses** [comment]
That sounds awesome -- thanks!

I'm subclassing `Layer` (`class TooltipLayer extends Layer { ... `), but for some reason the `instanceof` check fails... I don't know if it's a browser bug or a transpilation issue or something else -- I didn't delve too deeply into solving it, I just added the workaround.

**Comment by nicksrandall** [comment]
I have made these changes and added them to the documentation so I'm going to close this issue. Let me know if I've missed something. Thanks!
